README.md
# bitemporalpostgres

CHANGELOG.md
# Revision history for BitemporalPostgres

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
{-# LANGUAGE OverloadedStrings #-} 

  module Main where 
  import Prelude 
  import Database.PostgreSQL.Simple
      ( connectPostgreSQL )
  
  import MyPrelude
  import Data.Time.Clock 
  import Data.String(fromString)
  import Histories(createRow,History(..))
  import Versions(createRow, Version(..))
  import Validities(createRow, Validity(..))
  import Database.PostgreSQL.Simple.Time (Unbounded(Finite))
  
  import System.IO
  
  import Contracts -- (Contract(Contract), createEntityRow,ContractRevisionJSON)

  main :: IO ContractRevision
  main = do 
    conn <- connectPostgreSQL "host=localhost port=5432 user=postgres dbname=postgres connect_timeout=10"
    (histo:_)::[History] <- Histories.createRow conn (History nil "schrumpled8y") 
    print histo
    (version:_)::[Version] <- Versions.createRow conn (Version (0::Int64) (history_id histo) False)
    print version
    from <- getCurrentTime
    (_: _)::[Validity]<- Validities.createRow conn (Validity (0::Int64) (version_id version) (PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))(PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from)))))
    (contract: _)::[Contract] <- Contracts.createEntityRow conn (history_id histo)
    (contractrevision: _)::[ContractRevision] <- Contracts.createRevisionRow conn (ContractRevision (0::Int64) (contract_id contract) (PGRange(Inclusive(version_id version))(Exclusive(version_id version+1))) (ContractRevisionJSON "schnaps"))
    return contractrevision 

  createSchema :: IO (IO Int64)
  createSchema = do
    conn <- connectPostgreSQL "host=localhost port=5432 user=postgres dbname=postgres connect_timeout=10"
    handle <- openFile "schema.sql" ReadMode  
    qrytext <- hGetContents handle 
    _ <- execute_ conn "drop schema public cascade;"
    _ <- execute_ conn "create schema public;"
    return (execute_ conn (fromString qrytext))

src/Entities.hs
{-# LANGUAGE AllowAmbiguousTypes #-}

module Entities where
  import MyPrelude
  import Histories
  import Versions
  import Validities
  import Data.Time (getCurrentTime, addUTCTime)
  import Database.PostgreSQL.Simple.Time
  
  class Row r
    where
      get_id::r->Int64

  class (Show e , Row e, FromRow e) => EntityRow e
   where
    createEntityRow :: FromRow e => Connection -> UUID -> IO [e]
    set_history :: e -> UUID -> e

     --  (version:_)::[Version] <- Versions.createRow conn (Version (0::Int64) (history_id histo) False)
      --  print version
      --  from <- getCurrentTime
      --  (_: _)::[Validity]<- Validities.createRow conn (Validity (0::Int64) (version_id version) (PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))(PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from)))))
      --  (e: _)::[e] <- createEntityRow conn entity
  
  class (Show e , Row e,  FromRow e ) => RevisionRow e 
   where
    createRevisionRow :: FromRow e => Connection -> e -> IO [e]
    set_component :: e -> Int64 -> e
    
  class (Show e, EntityRow e, RevisionRow f) => BitemporalEntity e f
    where
      create:: Connection -> f  -> IO e
      create conn revision = do
         (h:_)::[History] <- Histories.createRow conn (History nil "schrumpled8y") 
         (version:_)::[Version] <- Versions.createRow conn (Version (0::Int64) (history_id h) False)
         from <- getCurrentTime
         (_: _)::[Validity]<- Validities.createRow conn (Validity (0::Int64) (version_id version) (PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))(PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))) 
         (created: _)::[e] <- createEntityRow conn (history_id h)
         _ <- createRevisionRow conn (set_component revision (get_id created))
         return created


src/Validities.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE EmptyDataDeriving #-}
{-# LANGUAGE MonoLocalBinds #-}

module Validities(createRow,Validity(..)) where
  import Prelude
  import MyPrelude(Action,Connection, FromRow,Generic,Int64,PGRange (..),query,ToRow, toField)
  import Database.PostgreSQL.Simple.Time (UTCTimestamp)
  
  data Validity = Validity {validity_id::Int64, ref_version::Int64, valid_ref::PGRange UTCTimestamp, valid_txn::PGRange UTCTimestamp }
    deriving (Show,Generic,FromRow)

  instance ToRow Validity where
  toRow :: Validity -> [Action]
  toRow a = [toField(ref_version a),toField (valid_ref a) , toField (valid_txn a)]

 
  createRow :: FromRow Validity => Connection -> Validity -> IO [Validity]
  createRow conn v = do
    query conn "insert into Validities (ref_version, valid_ref,valid_txn) values (?,?,?) returning id, ref_version, valid_ref,valid_txn" (toRow v)

src-internal/Contracts.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE EmptyDataDeriving #-}
{-# LANGUAGE DerivingVia #-}

{-# LANGUAGE MonoLocalBinds #-}

module Contracts (Contract(..),create, createEntityRow,ContractRevision(..),ContractRevisionJSON(..),createRevisionRow) where
   import MyPrelude(Connection,query, Generic,UUID, Int64, FromRow, ToRow,toRow, toField, ToField, FromField, Action, PGRange, nil)

   import Entities
   import Database.PostgreSQL.Simple.Newtypes (Aeson(..))
   import Data.Aeson
   import Database.PostgreSQL.Simple
   --import Histories

   data Contract = Contract {contract_id::Int64, ref_history::UUID}
    deriving (Show, Generic,FromRow)

   instance Row Contract where
      get_id :: Contract -> Int64
      get_id = contract_id

   instance ToRow Contract where
    toRow :: Contract -> [Action]
    toRow a = [toField (ref_history a)]


   instance EntityRow Contract
     where
      createEntityRow :: FromRow Contract => Connection -> UUID -> IO [Contract]
      createEntityRow conn refh = query conn "insert into contracts (ref_history) values (?) returning id, ref_history" (toRow (Contract{contract_id=0,ref_history=refh}))

      set_history :: Contract -> UUID -> Contract
      set_history c h =
       c{ref_history=h}

   data ContractRevisionJSON where
     ContractRevisionJSON :: {bubu :: String} -> ContractRevisionJSON
      deriving (Show, Generic)
      deriving (ToField, FromField) via Aeson ContractRevisionJSON   -- DerivingVia
   instance FromJSON ContractRevisionJSON
   instance ToJSON ContractRevisionJSON


   data ContractRevision = ContractRevision {contract_revision_id::Int64, ref_component::Int64, ref_valid::PGRange Int64, content:: ContractRevisionJSON}
    deriving (Show, Generic,FromRow)

   instance Row ContractRevision where
      get_id :: ContractRevision -> Int64
      get_id = contract_revision_id

   instance ToRow ContractRevision where
    toRow :: ContractRevision -> [Action]
    toRow a = [toField (ref_component a), toField (ref_valid a),toField (content a)]

   instance  RevisionRow ContractRevision
     where
      createRevisionRow :: FromRow ContractRevision => Connection -> ContractRevision -> IO [ContractRevision]
      createRevisionRow conn cr =
         query conn "insert into contract_revisions (ref_component,ref_valid,content) values (?,?,?) returning id, ref_component,ref_valid,content" (toRow cr)

      set_component :: ContractRevision -> Int64 -> ContractRevision
      set_component r cid =
         r{ref_component= cid}

   instance BitemporalEntity Contract ContractRevision


app/Main.hs
{-# LANGUAGE OverloadedStrings #-} 

  module Main where 
  import Prelude 
  import Database.PostgreSQL.Simple
      ( connectPostgreSQL )
  
  import MyPrelude
  import Data.Time.Clock 
  import Data.String(fromString)
  import Histories(createRow,History(..))
  import Versions(createRow, Version(..))
  import Validities(createRow, Validity(..))
  import Database.PostgreSQL.Simple.Time (Unbounded(Finite))
  
  import System.IO
  
  import Contracts -- (Contract(Contract), createEntity,createEntityRow,ContractRevisionJSON)

  main :: IO ContractRevision
  main = do 
    conn <- connectPostgreSQL "host=localhost port=5432 user=postgres dbname=postgres connect_timeout=10"
    (histo:_)::[History] <- Histories.createRow conn (History nil "schrumpled8y") 
    print histo
    (version:_)::[Version] <- Versions.createRow conn (Version (0::Int64) (history_id histo) False)
    print version
    from <- getCurrentTime
    (_: _)::[Validity]<- Validities.createRow conn (Validity (0::Int64) (version_id version) (PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))(PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from)))))
    (contract: _)::[Contract] <- Contracts.createEntityRow conn (history_id histo)
    (contractrevision: _)::[ContractRevision] <- Contracts.createRevisionRow conn (ContractRevision (0::Int64) (contract_id contract) (PGRange(Inclusive(version_id version))(Exclusive(version_id version+1))) (ContractRevisionJSON "schnaps"))
    return contractrevision
    
  newmain :: IO Contract
  newmain = do 
    conn <- connectPostgreSQL "host=localhost port=5432 user=postgres dbname=postgres connect_timeout=10"
    (histo:_)::[History] <- Histories.createRow conn (History nil "schrumpled8y") 
    print histo
    (version:_)::[Version] <- Versions.createRow conn (Version (0::Int64) (history_id histo) False)
    print version
    from <- getCurrentTime
    (_: _)::[Validity]<- Validities.createRow conn (Validity (0::Int64) (version_id version) (PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from))))(PGRange (Inclusive (Finite from))( Exclusive (Finite (addUTCTime (8*60*60) from)))))
    create conn (ContractRevision (0::Int64) (0::Int64) (PGRange(Inclusive(version_id version))(Exclusive(version_id version+1))) (ContractRevisionJSON "schnaps"))

  createSchema :: IO (IO Int64)
  createSchema = do
    conn <- connectPostgreSQL "host=localhost port=5432 user=postgres dbname=postgres connect_timeout=10"
    handle <- openFile "schema.sql" ReadMode  
    qrytext <- hGetContents handle 
    _ <- execute_ conn "drop schema public cascade;"
    _ <- execute_ conn "create schema public;"
    return (execute_ conn (fromString qrytext))